### PR TITLE
feat(PE-5540): recursive state traversal endpoint

### PR DESCRIPTION
**Pull request by dtfiedler**
This allows for traversing nested objects of contract states. It is a separate endpoint to make it clear it is traversing `state` of a contract. To avoid unbound recursion (specifically for unknown contracts that could have a gnarly state), we set the max depth of the path to 3 params.

Examples:

### observations/:epochStartHeight
```bash
curl http://localhost:3000/v1/contract/_NctcA2sRy1-J4OmIQZbYFPM17piNcbdBPH2ncX2RL8/state/observations/1353206 | jq '.'
{
  "contractTxId": "_NctcA2sRy1-J4OmIQZbYFPM17piNcbdBPH2ncX2RL8",
  "result": {
    "failureSummaries": {
      "NdZ3YRwMB2AMwwFYjKn1g88Y9nRybTo0qhS1ORq_E7g": [
        "M0WP8KSzCvKpzC-HPF1WcddLgGaL9J4DGi76iMnhrN4",
        "xN_aVln30LmoCffwmk5_kRkcyQZyZWy1o_TNtM_CTm0",
        "1H7WZIWhzwTH9FIcnuMqYkTsoyv1OTfGa_amvuYwrgo"
      ]
    },
    "reports": {
      "1H7WZIWhzwTH9FIcnuMqYkTsoyv1OTfGa_amvuYwrgo": "uDaaxpwivLuZR2KFDmcnlOIwyLc8a2DW2Uw6R60AeDY",
      "M0WP8KSzCvKpzC-HPF1WcddLgGaL9J4DGi76iMnhrN4": "10a3ZlqZTAbEl_t4dtkDFsiielkHSfM32G-0YUyeAr0",
      "xN_aVln30LmoCffwmk5_kRkcyQZyZWy1o_TNtM_CTm0": "4BRUUPDC6jhdCLfTtCs8BVr7Eam9uRccktQITVAgeac"
    }
  },
  "sortKey": "000001353765,0000000000000,b42b4698345a3e5499cd22df961d88c97bf731c63ddf4627b0183e924aefc4ee",
  "evaluationOptions": {
    "sourceType": "arweave"
  }
}

```
### observations/:epochStartHeight/reports
```bash
curl http://localhost:3000/v1/contract/_NctcA2sRy1-J4OmIQZbYFPM17piNcbdBPH2ncX2RL8/state/observations/1353206/reports | jq '.'
{
  "contractTxId": "_NctcA2sRy1-J4OmIQZbYFPM17piNcbdBPH2ncX2RL8",
  "result": {
    "1H7WZIWhzwTH9FIcnuMqYkTsoyv1OTfGa_amvuYwrgo": "uDaaxpwivLuZR2KFDmcnlOIwyLc8a2DW2Uw6R60AeDY",
    "M0WP8KSzCvKpzC-HPF1WcddLgGaL9J4DGi76iMnhrN4": "10a3ZlqZTAbEl_t4dtkDFsiielkHSfM32G-0YUyeAr0",
    "xN_aVln30LmoCffwmk5_kRkcyQZyZWy1o_TNtM_CTm0": "4BRUUPDC6jhdCLfTtCs8BVr7Eam9uRccktQITVAgeac"
  },
  "sortKey": "000001353765,0000000000000,b42b4698345a3e5499cd22df961d88c97bf731c63ddf4627b0183e924aefc4ee",
  "evaluationOptions": {
    "sourceType": "arweave"
  }
}

```
### gateways/:gatwayAddress
```bash
curl http://localhost:3000/v1/contract/_NctcA2sRy1-J4OmIQZbYFPM17piNcbdBPH2ncX2RL8/state/gateways/1H7WZIWhzwTH9FIcnuMqYkTsoyv1OTfGa_amvuYwrgo | jq '.'
{
  "contractTxId": "_NctcA2sRy1-J4OmIQZbYFPM17piNcbdBPH2ncX2RL8",
  "result": {
    "end": 0,
    "observerWallet": "1H7WZIWhzwTH9FIcnuMqYkTsoyv1OTfGa_amvuYwrgo",
    "operatorStake": 95048,
    "settings": {
      "fqdn": "permagate.io",
      "label": "Permagate",
      "note": "Owned and managed by @dtfiedler - if it breaks, find him.",
      "port": 443,
      "properties": "FH1aVetOoulPGqgYukj0VE0wIhDy90WiQoV3U2PeY44",
      "protocol": "https"
    },
    "start": 1252209,
    "stats": {
      "failedConsecutiveEpochs": 0,
      "passedEpochCount": 17,
      "submittedEpochCount": 2,
      "totalEpochParticipationCount": 25,
      "totalEpochsPrescribedCount": 21
    },
    "status": "joined",
    "vaults": {}
  },
  "sortKey": "000001353765,0000000000000,b42b4698345a3e5499cd22df961d88c97bf731c63ddf4627b0183e924aefc4ee",
  "evaluationOptions": {
    "sourceType": "arweave"
  }
}
```
### gateways/:gatwayAddress/stats
```bash
curl http://localhost:3000/v1/contract/_NctcA2sRy1-J4OmIQZbYFPM17piNcbdBPH2ncX2RL8/state/gateways/1H7WZIWhzwTH9FIcnuMqYkTsoyv1OTfGa_amvuYwrgo/stats | jq '.'
{
  "contractTxId": "_NctcA2sRy1-J4OmIQZbYFPM17piNcbdBPH2ncX2RL8",
  "result": {
    "failedConsecutiveEpochs": 0,
    "passedEpochCount": 17,
    "submittedEpochCount": 2,
    "totalEpochParticipationCount": 25,
    "totalEpochsPrescribedCount": 21
  },
  "sortKey": "000001353765,0000000000000,b42b4698345a3e5499cd22df961d88c97bf731c63ddf4627b0183e924aefc4ee",
  "evaluationOptions": {
    "sourceType": "arweave"
  }
}

```

### CODE REVIEW

File: docs/openapi.yaml [hunk 1]
```diff
@@ -319,12 +335,14 @@ paths:
 
   /contract/{contractTxId}/{field}:
     get:
-      summary: Returns the field in the contract state for a given transaction id
-      description: Returns the field in the contract state for a given transaction id
+      deprecated: true
+      summary: Returns the field in the contract state for a given transaction id. For traversing contract state, use /contract/{contractTxId}/state/{nestedPath}
+      description: Returns the field in the contract state for a given transaction id. For traversing contract state, use /contract/{contractTxId}/state/{nestedPath}
```
Comment (dtfiedler) [inline docs/openapi.yaml:340]: just marking our existing one layer deep handler as deprecated in favor of the new one

File: package.json [hunk 2]
```diff
@@ -17,7 +17,7 @@
     "start:prod": "yarn build && node dist/app.js",
     "build": "yarn clean && yarn && npx tsc --project ./tsconfig.json",
     "clean": "rimraf [ node_modules dist cache ]",
-    "test:docker": "docker compose up --exit-code-from test-runner --build",
+    "test:docker": "docker compose --env-file=.env.test up --exit-code-from test-runner --build",
```
Comment (dtfiedler) [inline package.json:20]: @kunstmusik - per your request!

File: src/utils.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import { MAX_PATH_DEPTH } from './constants';
+export function traverseObject<T>({
+  object,
+  path,
+}: {
+  object: any; // TODO: scope this to a type that can be traversed
```
Comment (fedellen) [inline src/utils.ts:22]: nit: handle any or eslint squelch

**Review by fedellen — APPROVED** [review]
love the tests 🤩

**Pull request merged**